xcube_zenodo/preload.py

```diff
@@ -0,0 +1,299 @@
+            self._cache_root, self._download_folder_name
+        )
+        self._events = [Event(data_id, np.nan) for data_id in data_ids]
+        self.lock = threading.Lock()
+        if preload_params.get("monitor_preload"):
+            for event in self._events:
+                event.subscribe(self._monitor_preload)
+        self._thread_download = None
+        self._thread_decompress = None
+        self._thread_prepare = None
+
+    @property
+    def is_cancelled(self) -> bool:
+        return self._is_cancelled
+
+    def cancel(self):
+        self._is_cancelled = True
+        self._thread_prepare.join()
+        self._thread_decompress.join()
+        self._thread_download.join()
+        for event in self._events:
+            event.update("Canceled", np.nan, "Preload has been canceled by user.")
+        self.close()
+
+    @property
+    def is_closed(self) -> bool:
+        return self._is_closed
+
+    def close(self):
+        self._is_closed = True
+        for event in self._events:
+            if event.status != "Preloaded":
+                if self._cache_store.has_data(event.data_id):
+                    self._cache_store.delete_data(event.data_id)
+                else:
+                    record, filename = event.data_id.split("/")
+                    format_id = estimate_file_format(event.data_id)
+                    dirname = filename.replace(f".{format_id}", "")
+                    data_id_mod = f"{record}/{dirname}"
+                    list_data_ids = self._cache_store.list_data_ids()
+                    list_data_ids_mod = [
+                        data_id for data_id in list_data_ids if data_id_mod in data_id
+                    ]
+                    for data_id in list_data_ids_mod:
+                        self._cache_store.delete_data(data_id)
+        if os.path.isdir(self._download_folder):
+            shutil.rmtree(self._download_folder)
+
+    def _monitor_preload(self):
+        rows = [
+            [
+                event.data_id,
+                event.status,
+                f"{event.progress * 100:.2f}%",
+                event.message,
+            ]
+            for event in self._events
+        ]
+        if is_jupyter():
+            table = tabulate.tabulate(
+                rows,
+                headers=["Data ID", "Status", "Progress", "Message"],
+                tablefmt="html",
+            )
+            IPython.display.clear_output(wait=True)
+            IPython.display.display(table)
+        else:
+            table = tabulate.tabulate(
+                rows,
+                headers=["Dataset", "Status", "Progress", "Message"],
+            )
+            os.system("clear" if os.name == "posix" else "cls")
+            print(table)
+
+    def preload_data(self, *data_ids: str, **preload_params):
+        self._download_data(*data_ids)
+        self._decompress_data(*data_ids)
+        self._prepare_data(*data_ids, **preload_params)
```

Comment:
QQ: We only have 3 threads in total, one for each of these tasks right? So each thread will do these tasks for all the data_ids. If that is the case, I do not see how these threads are useful here since they download the data one after the other, decompress after the download for each data_id one after other and so on. If you want to use concurrency to send mutliple download requests, the approach needs to be updated that each data_id has its own thread and so is not dependent on the other data_id status.
If this is not the case, I would be happy to discuss how this works.